File: frontend/src/services/authService.js
import { getApiBaseUrl, requestJson, requestWithAuthRetry } from "./apiClient";

export function saveSession({ user }) {
  if (user) {
    localStorage.setItem("current_user", JSON.stringify(user));
  }
}

export function clearSession() {
  localStorage.removeItem("current_user");
}

export function getStoredUser() {
  const rawUser = localStorage.getItem("current_user");
  if (!rawUser) return null;

  try {
    return JSON.parse(rawUser);
  } catch {
    return null;
  }
}

export function canAccessAdmin(user) {
  return Boolean(user?.can_access_admin);
}

export async function registerUser(payload) {
  return requestJson("/api/auth/register/", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
}

export async function loginUser(payload) {
  return requestJson("/api/auth/login/", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
}

export async function verifyEmail({ uid, token }) {
  return requestJson("/api/auth/verify-email/", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ uid, token }),
  });
}

export async function getCurrentUser() {
  const response = await requestWithAuthRetry("/api/auth/me/", {});

  if (!response.ok) {
    let data = null;
    try {
      data = await response.json();
    } catch {
      data = null;
    }

    const error = new Error(data?.detail || "Sesion invalida, vuelve a iniciar sesion.");
    error.status = response.status;
    throw error;
  }

  return response.json();
}

export async function requestPasswordReset(email) {
  return requestJson("/api/auth/password-reset/request/", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ email }),
  });
}

export async function confirmPasswordReset({ uid, token, password, passwordConfirm }) {
  return requestJson("/api/auth/password-reset/confirm/", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ uid, token, password, password_confirm: passwordConfirm }),
  });
}

export async function logoutUser() {
  return requestJson("/api/auth/logout/", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({}),
  });
}

export { getApiBaseUrl };

File: mobile/lib/features/auth/data/auth_service.dart
import '../../../core/network/api_client.dart';
import 'models/auth_user.dart';

class AuthService {
  AuthService({ApiClient? apiClient}) : _apiClient = apiClient ?? ApiClient();

  final ApiClient _apiClient;

  Future<AuthSession> login({
    required String email,
    required String password,
  }) async {
    final response = await _apiClient.post(
      '/api/auth/login/',
      body: {'email': email.trim(), 'password': password},
    );

    final data = _apiClient.parseJsonMap(response);

    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw AuthServiceException(
        _extractDetail(data, fallback: 'No se pudo iniciar sesion.'),
      );
    }

    final access = (data['access'] as String? ?? '').trim();
    final refresh = (data['refresh'] as String? ?? '').trim();
    final userJson = data['user'];

    if (access.isEmpty ||
        refresh.isEmpty ||
        userJson is! Map<String, dynamic>) {
      throw const AuthServiceException('Respuesta invalida del servidor.');
    }

    return AuthSession(
      accessToken: access,
      refreshToken: refresh,
      user: AuthUser.fromJson(userJson),
    );
  }

  Future<void> register({
    required String firstName,
    required String lastName,
    required String email,
    required String password,
  }) async {
    final response = await _apiClient.post(
      '/api/auth/register/',
      body: {
        'first_name': firstName.trim(),
        'last_name': lastName.trim(),
        'email': email.trim(),
        'password': password,
      },
    );

    final data = _apiClient.parseJsonMap(response);
    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw AuthServiceException(
        _extractDetail(data, fallback: 'No se pudo completar el registro.'),
      );
    }
  }

  Future<String> requestPasswordReset(String email) async {
    final response = await _apiClient.post(
      '/api/auth/password-reset/request/',
      body: {'email': email.trim()},
    );

    final data = _apiClient.parseJsonMap(response);
    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw AuthServiceException(
        _extractDetail(
          data,
          fallback: 'No se pudo procesar la solicitud de recuperacion.',
        ),
      );
    }

    final detail = data['detail'];
    if (detail is String && detail.trim().isNotEmpty) {
      return detail.trim();
    }

    return 'Si el correo existe, te enviaremos un enlace de recuperacion.';
  }

  Future<String> confirmPasswordReset({
    required String uid,
    required String token,
    required String password,
    required String passwordConfirm,
  }) async {
    final response = await _apiClient.post(
      '/api/auth/password-reset/confirm/',
      body: {
        'uid': uid.trim(),
        'token': token.trim(),
        'password': password,
        'password_confirm': passwordConfirm,
      },
    );

    final data = _apiClient.parseJsonMap(response);
    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw AuthServiceException(
        _extractDetail(data, fallback: 'No se pudo actualizar la contrasena.'),
      );
    }

    final detail = data['detail'];
    if (detail is String && detail.trim().isNotEmpty) {
      return detail.trim();
    }

    return 'Contrasena actualizada correctamente.';
  }

  Future<String> verifyEmail({
    required String uid,
    required String token,
  }) async {
    final response = await _apiClient.post(
      '/api/auth/verify-email/',
      body: {'uid': uid.trim(), 'token': token.trim()},
    );

    final data = _apiClient.parseJsonMap(response);
    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw AuthServiceException(
        _extractDetail(data, fallback: 'No se pudo verificar el correo.'),
      );
    }

    final detail = data['detail'];
    if (detail is String && detail.trim().isNotEmpty) {
      return detail.trim();
    }

    return 'Correo verificado correctamente. Ya puedes iniciar sesion.';
  }

  Future<AuthUser> getProfile(String accessToken) async {
    final response = await _apiClient.get(
      '/api/auth/me/',
      headers: {'Authorization': 'Bearer $accessToken'},
    );

    final data = _apiClient.parseJsonMap(response);
    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw AuthServiceException(
        _extractDetail(data, fallback: 'No se pudo recuperar el perfil.'),
      );
    }

    return AuthUser.fromJson(data);
  }

  Future<String> refreshToken(String refreshToken) async {
    final response = await _apiClient.post(
      '/api/auth/refresh/',
      body: {'refresh': refreshToken},
    );

    final data = _apiClient.parseJsonMap(response);
    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw AuthServiceException(
        _extractDetail(data, fallback: 'No se pudo renovar la sesion.'),
      );
    }

    final access = (data['access'] as String? ?? '').trim();
    if (access.isEmpty) {
      throw const AuthServiceException(
        'No se recibio access token al renovar sesion.',
      );
    }

    return access;
  }

  Future<void> logout({required String accessToken}) async {
    await _apiClient.post(
      '/api/auth/logout/',
      headers: {'Authorization': 'Bearer $accessToken'},
    );
  }

  String _extractDetail(Map<String, dynamic> data, {required String fallback}) {
    final detail = data['detail'];
    if (detail is String && detail.trim().isNotEmpty) {
      return detail.trim();
    }

    final errors = data.entries.where((entry) => entry.value is List).toList();
    if (errors.isNotEmpty) {
      final first = errors.first.value as List<dynamic>;
      if (first.isNotEmpty) {
        return first.first.toString();
      }
    }

    return fallback;
  }
}

class AuthServiceException implements Exception {
  const AuthServiceException(this.message);

  final String message;

  @override
  String toString() => message;
}
